src/containers/frequency_map.rs

```diff
@@ -0,0 +1,1410 @@
+    {
+        let r = self.map.remove_entry(key);
+
+        if let Some((_, v)) = &r {
+            self.total -= v;
+        };
+
+        r
+    }
+
+    /// Reserves capacity for at least `additional` more elements to be
+    /// inserted in the instance.
+    #[inline]
+    pub fn reserve(&mut self, additional: usize) {
+        self.map.reserve(additional);
+    }
+
+    /// Shrinks the capacity of the map with a lower limit.
+    #[inline]
+    pub fn shrink_to(&mut self, min_capacity: usize) {
+        self.map.shrink_to(min_capacity)
+    }
+
+    /// Shrinks the capacity of the map as much as possible.
+    #[inline]
+    pub fn shrink_to_fit(&mut self) {
+        self.map.shrink_to_fit()
+    }
+}
+
+impl<K> FrequencyMap<K> {
+    /// Retains only the elements specified by the predicate.
+    #[inline]
+    pub fn retain<F>(&mut self, f: F)
+    where
+        F: Fn(&K, isize) -> bool,
+    {
+        let mut dropped_total = 0;
+
+        self.map.retain(|key, count| {
+            if f(key, *count) {
+                true
+            } else {
+                dropped_total += *count;
+
+                false
+            }
+        });
+
+        self.total -= dropped_total;
+    }
+}
+
+// Non-mutating methods
+
+impl<K : std_cmp::Eq + std_hash::Hash> FrequencyMap<K> {
+    /// Indicates whether a record exists for the given key.
+    #[inline]
+    pub fn contains_key<Q>(&self, key: &Q) -> bool
+    where
+        K: std_borrow::Borrow<Q>,
+        Q: std_hash::Hash + std_cmp::Eq + ?Sized,
+    {
+        self.map.contains_key(key)
+    }
+
+    /// Obtains the count corresponding to the given key, obtaining 0 in the
+    /// case that no such record exists.
+    #[inline]
+    pub fn get<Q>(&self, key: &Q) -> isize
+    where
+        K: std_borrow::Borrow<Q>,
+        Q: std_hash::Hash + std_cmp::Eq + ?Sized,
+    {
+        *self.get_(key)
+    }
+}
+
+impl<K> FrequencyMap<K> {
+    /// An iterator visiting all key-count pairs in arbitrary order. The
+    /// iterator element type is `(&'a K, &'a isize)`.
+    #[inline]
+    pub fn iter(&self) -> std::collections::hash_map::Iter<'_, K, isize> {
+        self.map.iter()
+    }
+}
+
+impl<K> FrequencyMap<K> {
+    /// Returns the number of elements the map can hold without
+    /// reallocation.
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.map.capacity()
+    }
+}
+
+impl<K> FrequencyMap<K> {
+    /// Indicates whether the instance contains no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.is_empty_()
+    }
+
+    /// Indicates the number of records.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len_()
+    }
+
+    /// Indicates the total count across all keys.
+    #[inline]
+    pub fn total(&self) -> isize {
+        self.total_()
+    }
+}
+
+// Implemetation
+
+impl<K : std_cmp::Eq + std_hash::Hash> FrequencyMap<K> {
+    fn get_<Q>(&self, key: &Q) -> &isize
```

Comment:
Why does this return a reference?  Isn't it easier to return the value itself?

Reply:
Have been able to do this with recent `UnicodePointMap`, so will come back to this and see if same can be achieved.

Reply:
I misspoke. Albeit some of its public methods can return by value, `UnicodePointMap`'s `#get_()` also has to return by ref because subscript/index operator must do so.